### PR TITLE
MM-36687 CRT push notifications

### PR DESCRIPTION
**Pull request by anurag6713**
#### Summary
* Sends `is_crt_enabled` in the push notification to handle notifications on the client side and group them accordingly
* Sends correct badge count when CRT is enabled
* `root_id` is always sent for all the notifications
* `team_id` is sent

### Related PRs
https://github.com/mattermost/mattermost-mobile/pull/5669
https://github.com/mattermost/mattermost-push-proxy/pull/85
https://github.com/mattermost/enterprise/pull/1096

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-36687

#### Release Note
```
NONE
```

### CODE REVIEW

File: app/notification_push.go [hunk 1]
```diff
@@ -217,31 +218,45 @@ func (a *App) getPushNotificationMessage(contentsConfig, postMessage string, exp
...
+		unreadTeamsList, err := a.GetTeamsUnreadForUser("", userID, true)
+		if err != nil {
+			return model.NewAppError("clearPushNotificationSync", "app.user.get_unread_count.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+		msg.Badge = 0
+		for i := 0; i < len(unreadTeamsList); i++ {
+			team := unreadTeamsList[i]
+			msg.Badge += int(team.MentionCountRoot)
+			msg.Badge += int(team.ThreadMentionCount)
+		}
```
Comment (agnivade) [inline app/notification_push.go:241]: More idiomatically written as:

```go
for _, team := range unreadTeamsList {
  msg.Badge += int(team.MentionCountRoot) + int(team.ThreadMentionCount)
}
```

File: app/notification_push.go [hunk 2]
```diff
@@ -217,31 +218,45 @@ func (a *App) getPushNotificationMessage(contentsConfig, postMessage string, exp
...
 		ChannelId:        channelID,
+		RootId:           rootID,
 		ContentAvailable: 1,
 	}
 
-	unreadCount, err := a.Srv().Store.User().GetUnreadCount(userID)
-	if err != nil {
-		return model.NewAppError("clearPushNotificationSync", "app.user.get_unread_count.app_error", nil, err.Error(), http.StatusInternalServerError)
+	msg.IsCRTEnabled = a.isCRTEnabledForUser(userID)
+	if msg.IsCRTEnabled {
```
Comment (agnivade) [inline app/notification_push.go:231]: Let's add a test case which tests `IsCRTEnabled` true and verifies the badge count.

File: api4/system.go [hunk 3]
```diff
@@ -513,11 +513,6 @@ func pushNotificationAck(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	if _, appErr := c.App.GetPostIfAuthorized(ack.PostId, c.AppContext.Session()); appErr != nil {
```
Comment (agnivade) [inline api4/system.go:516]: It is not clear to me why we moved this inside the `ack.PostId != ""` condition. Can you clarify?

**Review by agnivade — COMMENTED** [review]
Thanks Anurag. Got a few minor comments, but mainly needs some unit tests around the new functionality.

File: api4/system.go (hunk 3 above)
Comment (anurag6713) [reply api4/system.go:516]: While we clear channel & thread notifications, we do not pass PostId. While we are trying to ack those `clear notifications` .. they are failing. As we may or may not get the PostId from the client, I have moved this condition to post notification acknowledgement.

**Review by agnivade — APPROVED** [review]
Nice!

**Pull request merged**